utils/test_results.py

```diff
@@ -1,101 +1,315 @@
-                        distinct=True,
-                        filter=Q(outcome__in=["failure", "error"]),
-                    )
-                ),
-                0,
-                output_field=IntegerField(),
-            ),
-            avg_duration=Avg("duration_seconds"),
-            name=F("test__name"),
+                + Cast(Sum(F("fail_count")), output_field=FloatField())
+            )
+        ),
+        updated_at=Max("latest_run"),
+        commits_where_fail=ArrayLength(Array(Subquery(commits_where_fail_sq))),
+        last_duration=Subquery(latest_duration_sq),
+        avg_duration=Avg("avg_duration_seconds"),
+        name=F("test__name"),
+    )
+
+    return aggregation_of_test_results
+
+
+def percent_diff(a: int | float, b: int | float) -> int | float:
+    c = (a - b) / b * 100
+    return c
+
+
+def helper(
+    ls: list[str],
+    curr_numbers: dict[str, int | float],
+    past_numbers: dict[str, int | float],
+) -> dict[str, int | float]:
+    diff_dict = {}
+
+    for s in ls:
+        if past_numbers.get(s):
+            diff_dict[f"{s}_percent_change"] = percent_diff(
+                curr_numbers[s], past_numbers[s]
+            )
+
+    return diff_dict
+
+
+def get_test_results_aggregate_numbers(
+    repo: Repository, since: dt.datetime, until: dt.datetime | None = None
+) -> dict[str, float | int]:
+    totals = DailyTestRollup.objects.filter(
+        repoid=repo.repoid, date__gt=since, branch=repo.branch
+    )
+
+    if until:
+        totals = totals.filter(date__lte=until)
+
+    num_tests = totals.distinct("test_id").count()
+
+    slowest_test_ids = (
+        totals.values("test")
+        .annotate(
+            runtime=Avg("avg_duration_seconds")
+            * (Sum("pass_count") + Sum("fail_count"))
         )
-        .values(
-            "failure_rate",
-            "commits_where_fail",
-            "avg_duration",
-            "name",
-            "updated_at",
+        .order_by("-runtime")
+        .values("test_id")[0 : slow_test_threshold(num_tests)]
+    )
+
+    slowest_tests_duration = (
+        totals.filter(test_id__in=slowest_test_ids)
+        .values("repoid")
+        .annotate(
+            slowest_tests=Avg("avg_duration_seconds")
+            * (Sum("pass_count") + Sum("fail_count"))
         )
+        .values("slowest_tests")
+    )
+
+    test_headers = totals.values("repoid").annotate(
+        total_run_time=Avg("avg_duration_seconds")
+        * (Sum("pass_count") + Sum("fail_count")),
+        slowest_tests_duration=slowest_tests_duration,
+        skips=Sum("skip_count"),
+        fails=Sum("fail_count"),
+    )
+
+    return test_headers[0] if len(test_headers) > 0 else {}
+
+
+def generate_test_results_aggregates(
+    repoid: int, history: dt.timedelta | None = None
+) -> dict[str, float | int] | None:
+    repo = Repository.objects.get(repoid=repoid)
+    time_ago = (
+        (dt.datetime.now(dt.UTC) - history) if history is not None else thirty_days_ago
```

Comment:
we can simplify these if else's a bit if we set History to be a const dependent on it's value being passed in OR dt.timedelta(days=30)

so then we can just do  time_ago = (
        (dt.datetime.now(dt.UTC) - history) and  double_time_ago = time_ago - 
        history
 
WIth something like history = history or dt.timedelta(days=30)